src/vsr/replica_test.zig

```diff
@@ -0,0 +1,920 @@
+        const paths = t.peer_paths(peer, direction);
+        for (paths.constSlice()) |path| {
+            t.cluster.network.link_filter(path).* = LinkFilter.initFull();
+        }
+    }
+
+    pub fn drop_all(t: *const TestReplicas, peer: ProcessSelector, direction: LinkDirection) void {
+        const paths = t.peer_paths(peer, direction);
+        for (paths.constSlice()) |path| t.cluster.network.link_filter(path).* = LinkFilter{};
+    }
+
+    pub fn pass(
+        t: *const TestReplicas,
+        peer: ProcessSelector,
+        direction: LinkDirection,
+        command: vsr.Command,
+    ) void {
+        const paths = t.peer_paths(peer, direction);
+        for (paths.constSlice()) |path| t.cluster.network.link_filter(path).insert(command);
+    }
+
+    pub fn drop(
+        t: *const TestReplicas,
+        peer: ProcessSelector,
+        direction: LinkDirection,
+        command: vsr.Command,
+    ) void {
+        const paths = t.peer_paths(peer, direction);
+        for (paths.constSlice()) |path| t.cluster.network.link_filter(path).remove(command);
+    }
+
+    const paths_max = constants.nodes_max * (constants.nodes_max + constants.clients_max);
+
+    fn peer_paths(
+        t: *const TestReplicas,
+        peer: ProcessSelector,
+        direction: LinkDirection,
+    ) std.BoundedArray(Network.Path, paths_max) {
+        var paths = std.BoundedArray(Network.Path, paths_max){ .buffer = undefined };
+        const peers = t.context.processes(peer);
+        for (t.replicas.constSlice()) |a| {
+            const process_a = Process{ .replica = a };
+            // TODO Zig: This should iterate over values instead of pointers once the miscompilation
+            // segfault is fixed.
+            for (peers.constSlice()) |*process_b| {
+                if (direction == .bidirectional or direction == .outgoing) {
+                    paths.appendAssumeCapacity(.{ .source = process_a, .target = process_b.* });
+                }
+                if (direction == .bidirectional or direction == .incoming) {
+                    paths.appendAssumeCapacity(.{ .source = process_b.*, .target = process_a });
+                }
+            }
+        }
+        return paths;
+    }
+};
+
+const TestClients = struct {
+    context: *TestContext,
+    cluster: *Cluster,
+    clients: std.BoundedArray(usize, constants.clients_max),
+    requests: usize = 0,
+
+    pub fn request(t: *TestClients, requests: usize) usize {
+        if (t.requests < requests) {
+            var requests_remaining = requests - t.requests;
+            while (requests_remaining > 0) {
+                for (t.clients.constSlice()) |c| {
+                    t.context.client_requests[c] += 1;
+                    requests_remaining -= 1;
+                    if (requests_remaining == 0) break;
+                }
+            }
+            t.requests = requests;
+        }
```

Comment:
It feels like this could be simplified a bit, perhaps to 

```zig
    pub fn request(t: *TestClients, requests: usize) usize {
        assert(t.requests <= requests);
        defer assert(t.requests == requests);

        outer: while (true) {
            for (t.clients.constSlice()) |c| {
                if (t.requests == requests) break :outer;
                t.context.client_requests[c] += 1;
                t.requests += 1;
            }
        }
```